src/foam/dao/AdapterDAO.js

```diff
@@ -38,15 +52,37 @@ foam.CLASS({
           documentation: `"adapt(o)" adapts input to type expected by
               "delegate" sink.`,
           required: true
+        },
+        {
+          class: 'FObjectProperty',
+          of: 'foam.dao.Adapter',
+          name: 'adapter',
+          required: true
         }
       ],
 
       methods: [
-        function put(o, sub) {
-          this.delegate.put(this.adapt(o), sub);
+        {
+          name: 'put',
+          code: function put(o, sub) {
+            this.delegate.put(this.adapt(o), sub);
+          },
+          javaCode: `
+            FObject innerObject = (FObject) obj;
+            FObject outerObject = getAdapter().adaptFromDelegate(getX(), innerObject);
```

Comment:
Like this JS version, I would just call adapt(), instead of getAdapter().adaptFromDelegate() and then remove the 'adapter' property and have the adapt and unadapt methods just do what you want by default. Behaviour can be modified if needed by just subclassing and overriding those two methods.